compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1181,13 +1181,37 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] w
    *  for equality would give the wrong result, so we should not use the sets
    *  for comparisons.
    */
-  def canCompare(atoms: Set[Type]): Boolean =
-    ctx.phase.isTyper || {
+  def compareAtoms(tp1: Type, tp2: Type): Option[Boolean] =
```

Comment:
The use of an Option here might noticeably increase our allocation rate, a tri-state enum could be used instead.

Reply:
I don't think so, since the overwhelming majority of calls should give `None`. You get a `Some` only if both compared sides are consist of atoms.